### ISSUE TITLE
503s and `_No_route_to_host` errors due to routing to non-existent Endpoints

### ISSUE THREAD

**Issue by coro** [envoyproxy/gateway#4685]
*Description*:
We have been seeing many 503 errors when connecting to a Service with a lot of Pod churn.

We also saw in the logs that in these cases, the `upstream_host` that Envoy was attempting to connect to was for Pods that no longer existed in the cluster. These Pods could have been terminated over 50 mins earlier.

*Repro steps*:
Simple setup of Gateway (AWS NLB) -> HTTPRoute -> Service pointing to a Deployment with a lot of Pod churn.

*Environment*:
Gateway: v1.2.1 (also seen on v1.1.3) (**not** seen on v1.1.1)
Envoy: v1.32.1 (also seen on v1.31.1) (**not** seen on v1.31.0)
EKS cluster v1.29

*Logs*:
```json
{
  "start_time": "2024-11-08T13:54:35.382Z",
  "method": "POST",
  "x-envoy-origin-path": "/v1/models/custom-model:predict",
  "protocol": "HTTP/1.1",
  "response_code": "503",
  "response_flags": "UF",
  "response_code_details": "upstream_reset_before_response_started{remote_connection_failure|delayed_connect_error:_No_route_to_host}",
  "connection_termination_details": "-",
  "upstream_transport_failure_reason": "delayed_connect_error:_No_route_to_host",
  "bytes_received": "494",
  "bytes_sent": "165",
  "duration": "3060",
  "x-envoy-upstream-service-time": "-",
  "x-forwarded-for": "10.0.99.64",
  "user-agent": "python-requests/2.32.3",
  "x-request-id": "3dbc6bcc-79b1-4ec1-855c-1569b69f5416",
  ":authority": "kserve-detection.prod.signal",
  "upstream_host": "10.0.101.243:8080",
  "upstream_cluster": "httproute/default/detection/rule/0",
  "upstream_local_address": "-",
  "downstream_local_address": "10.0.101.92:10080",
  "downstream_remote_address": "10.0.99.64:41562",
  "requested_server_name": "-",
  "route_name": "httproute/default/detection/rule/0/match/0/kserve-detection_prod_signal"
}
```

The generated cluster config for this route was:

```yaml
   "dynamic_active_clusters": [
    {
     "version_info": "a998a3b0100c28a29ff3ee662cb01dc45463f0260c9a6a50107dcc12a45938fc",
     "cluster": {
      "@type": "type.googleapis.com/envoy.config.cluster.v3.Cluster",
      "name": "httproute/default/detection/rule/0",
      "type": "EDS",
      "eds_cluster_config": {
       "eds_config": {
        "ads": {},
        "resource_api_version": "V3"
       },
       "service_name": "httproute/default/detection/rule/0"
      },
      "connect_timeout": "10s",
      "per_connection_buffer_limit_bytes": 32768,
      "lb_policy": "LEAST_REQUEST",
      "circuit_breakers": {
       "thresholds": [
        {
         "max_retries": 1024
        }
       ]
      },
      "dns_lookup_family": "V4_ONLY",
      "outlier_detection": {},
      "common_lb_config": {
       "locality_weighted_lb_config": {}
      },
      "ignore_health_on_host_removal": true
     },
     "last_updated": "2024-11-08T14:29:21.557Z"
    },
...
```

**Comment by arkodg** [comment]
hey @coro, we haven't seen such issues yet, can you also run `kubectl get endpointslice -l <service label> -o yaml` as well as `egctl config envoy-proxy endpoint` multiple times over a few seconds, so we can better understand the consistency issue

to avoid this, you can set up retries and passive health checks, here's an example
```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: BackendTrafficPolicy
metadata:
  name: retry-for-route
spec:
  targetRefs:
    - group: gateway.networking.k8s.io
      kind: HTTPRoute
      name: backend
  retry:
    numRetries: 5
    perRetry:
      backOff:
        baseInterval: 100ms
        maxInterval: 10s
      timeout: 250ms
    retryOn:
      httpStatusCodes:
        - 500
      triggers:
        - connect-failure
        - retriable-status-codes
  healthCheck:
    passive:
      baseEjectionTime: 10s
      interval: 2s
      maxEjectionPercent: 100
      consecutive5XxErrors: 1 
      consecutiveGatewayErrors: 0
      consecutiveLocalOriginFailures: 1
      splitExternalLocalOriginErrors: false
```

**Comment by arkodg** [comment]
cc @dprotaso

**Comment by coro** [comment]
@arkodg Just a quick update, we can confirm that this does not happen with EG v1.1.1 / Envoy v1.31.0. Rolling back seems to have resolved this issue. We haven't had a moment to safely reproduce this and get your additional debug info just yet, so we will let you know again when we have that.

**Comment by coro** [comment]
cc @evilr00t @sam-burrell

**Comment by sam-burrell** [comment]
We can confirm that pods can be terminated and `egctl config envoy-proxy endpoint` for the route confirms the pod is not in the list of endpoints `lbEndpoints`, there is no mention of the ip of the pod (healthy or otherwise) 

Yet we still get logs after **that point in time** from envoy saying `upstream_reset_before_response_started{remote_connection_failure|delayed_connect_error:_Connection_refused}` for the upstream ip of a pod not in the list of `lbEndpoints` from envoy.

There seems to be some routing still within envoy that is trying to route to a dead/non existant pod. Perhaps this is cached or perhaps there is a long lived connection still open or trying to be used?

This is continuing from the same environment as @coro  (We work on the same team)

**Comment by evilr00t** [comment]
Hi! 

Looks like we were able to identify the issue.

We've started comparing proxy endpoints `egctl config envoy-proxy endpoint -n kube-system envoy_pod` 
and since v1.1.3 (we've checked 1.1.1 and 1.1.2) there seems to be a regression.

After upgrade controller does not update it's gatewayclasses endpoints anymore - only during the startup which as you can imagine in very dynamic k8s world causes a lot of issues. 

We were comparing different tags and I think this is what caused the regression altho I'm not 100% sure - just giving you a hint.

https://github.com/envoyproxy/gateway/issues/4336
https://github.com/envoyproxy/gateway/pull/4337/files

**Comment by arkodg** [comment]
unsure how status updater can affect endpointslice reconciliation
cc @zhaohuabing

**Comment by ovaldi** [comment]
We face same issue too.

**Comment by ligol** [comment]
We're getting the same error on our side too, rollbacking to v1.1.2 seemed fixing the issue on our side too

**Comment by zhaohuabing** [comment]
@ligol @ovaldi @evilr00t @sam-burrell @coro Can you try 
v1.1.3+ envoyproxy/envoy:distroless-v1.31.2 
or
v1.1.2 envoyproxy/envoy:distroless-v1.31.3? 
 
 Looks like the only significant change within v1.1.3 is the upgrade of envoy to v1.31.3.

**Comment by ligol** [comment]
I will try to test the different combinaison tomorrow, when I'll have some traffic on the preprod environement, otherwise I don't really have any other way to reproduce

**Comment by sam-burrell** [comment]
These are all the combinations we have tested

**Working**
gateway 1.1.1
Envoy image envoyproxy/envoy:distroless-v1.31.1

**Not Working**
gateway 1.1.3
Envoy image envoyproxy/envoy:distroless-v1.31.1

Not Working
gateway 1.1.2
Envoy image envoyproxy/envoy:distroless-v1.31.2

Not Working
gateway 1.1.3
Envoy image envoyproxy/envoy:distroless-v1.31.2

**Working**
gateway 1.1.2
Envoy image envoyproxy/envoy:distroless-v1.31.3

**Not Working**
gateway 1.2.1
Envoy image envoyproxy/envoy:distroless-v1.32.1

**Comment by sam-burrell** [comment]
We have turned `debug` logs on for `envoy-gateway-controller` and see nothing obvious.
We do not see any logs in `envoy-gateway-controller` matching `"error|warning|fatal|Error"`
We do not see any logs in `envoy-gateway-pods` other than the `503s`

**Comment by sam-burrell** [comment]
Is there anything we can do to assist with this?

**Comment by zhaohuabing** [comment]
@sam-burrell @ovaldi I’m working on reproducing this in my test environment (Kind cluster) for further debugging. Have you found any reliable way to reproduce it?”

**Comment by zhaohuabing** [comment]
I tried the below setup with both 1.2.1 and 1.1.2 

Constantly create and delete pods with:
```
while true
do
kubectl patch deployment backend -p '{"spec":{"replicas":100}}'
sleep 10
kubectl patch deployment backend -p '{"spec":{"replicas":1}}'
sleep 10
done
```
And access the Gateway with 

```
while true
do
status_code=`curl -o /dev/null -s -w "%{http_code}\n"  -HHost:www.example.com http://172.18.0.200`
if [ "$status_code" -eq 503 ]; then
  echo "503"
fi
done
```

And I got some 503 erros for both the 1.2.1 and 1.1.2, after stopping creating and deleting pods, the 503 errors stopped for both versions.

**Comment by zhaohuabing** [comment]
> The following shell output is how we are defining **Not Working**
> We get results like the following after randomly killing a few pods

I think it's acceptable for the endpoints in envoy xDS temporarily differ from the k8s endpointslice. This delay occurs because Envoy Gateway (EG) needs time to propagate the changes to Envoy. However, if an endpoint still exists in Envoy for a while after being deleted from the EndpointSlice, it indicates a problem.

**Comment by zhaohuabing** [comment]
@ligol @ovaldi @evilr00t @sam-burrell @coro Could you please try if https://github.com/envoyproxy/gateway/pull/4754 can fix this?

**Comment by dromadaire54** [comment]
@zhaohuabing we are going to try if the revert fix the issue.

**Comment by zhaohuabing** [comment]
> @zhaohuabing we are going to try if the revert fix the issue.

Please also try https://github.com/envoyproxy/gateway/pull/4767 if you have a stable reproducable env. Thanks!

**Comment by evilr00t** [comment]
Just checked the latest v0.0.0-latest and this bug still happens - only 1 of the proxies is being updates during the scale-up / scale-down. 

Screenshot included:

![Screenshot 2024-11-22 at 10 34 59 AM](https://github.com/user-attachments/assets/d9673a55-6dcc-41bb-8e80-1b06054864a5)

**Comment by arkodg** [comment]
@zhaohuabing can you provide a custom image from your docker hub repo (with the commit reverted) that others can try ?

**Comment by zhaohuabing** [comment]
*  zhaohuabing/gateway-dev:8c9d33e04  https://github.com/envoyproxy/gateway/pull/4767


* zhaohuabing/gateway-dev:d3e3c6bf7  https://github.com/envoyproxy/gateway/pull/4754

**Comment by zhaohuabing** [comment]
Update, please use zhaohuabing/gateway-dev:e406088d7 for https://github.com/envoyproxy/gateway/pull/4754

**Comment by zhaohuabing** [comment]
> Just checked the latest v0.0.0-latest and this bug still happens - only 1 of the proxies is being updates during the scale-up / scale-down.
> 
> Screenshot included:
> 
> ![Screenshot 2024-11-22 at 10 34 59 AM](https://private-user-images.githubusercontent.com/12815767/388903217-d9673a55-6dcc-41bb-8e80-1b06054864a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NTgwNzYsIm5iZiI6MTczMjQ1Nzc3NiwicGF0aCI6Ii8xMjgxNTc2Ny8zODg5MDMyMTctZDk2NzNhNTUtNmRjYy00MWJiLThlODAtMWIwNjA1NDg2NGE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI0VDE0MTYxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYxYWI2ODQzYzcyYmVjMjIxZjY3YzZiOTIyMmQwMmUyNDNhNjFmYjk3ZDIxNDY2Mjk3MDQ5M2E2NTRkMGNiM2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.EaxE2vhu2E9YVLgZxdnLkqcwXKhRAcWnY4crNN1rsJI)

@evilr00t  It's temporary or the deleted pod IPs will be in the endpoints for a long time?

**Comment by evilr00t** [comment]
> > Just checked the latest v0.0.0-latest and this bug still happens - only 1 of the proxies is being updates during the scale-up / scale-down.
> > Screenshot included:
> > ![Screenshot 2024-11-22 at 10 34 59 AM](https://private-user-images.githubusercontent.com/12815767/388903217-d9673a55-6dcc-41bb-8e80-1b06054864a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NTgwNzYsIm5iZiI6MTczMjQ1Nzc3NiwicGF0aCI6Ii8xMjgxNTc2Ny8zODg5MDMyMTctZDk2NzNhNTUtNmRjYy00MWJiLThlODAtMWIwNjA1NDg2NGE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI0VDE0MTYxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYxYWI2ODQzYzcyYmVjMjIxZjY3YzZiOTIyMmQwMmUyNDNhNjFmYjk3ZDIxNDY2Mjk3MDQ5M2E2NTRkMGNiM2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.EaxE2vhu2E9YVLgZxdnLkqcwXKhRAcWnY4crNN1rsJI)
> 
> @evilr00t It's temporary or the deleted pod IPs will be in the endpoints for a long time?

on provided screenshot, when I've scaled up the replica set - new IPs of Pods were added only to 1 envoy proxy. Deleted pods are removed from that 1 working proxy. 2 Remaining proxies are not updated at all and shows the IPs of Pods that were available during the startup of those proxies.

**Comment by tskinner-oppfi** [comment]
Thanks for looking into this @zhaohuabing  This is a great project!

I am seeing similar issues and finally took some time to try and make some reproducible steps using some of the scripts in the gateway repo and here is what I've come up with:

1. Spin up a local test cluster with eg and the demo app (I didn't look around the repo too well so if there is a better way, please tell)
	 ```
	make create-cluster
	make kube-install-image
	make kube-deploy
	make kube-demo
	 ```
2. Scale up the envoy proxies and gateways to 3. The issue doesn't seem to happen with only a single replica.
3. Scale the backend (demo app) up to 10 replicas or something
4. Add a listener to the gateway. I have something similar listener in my cluster and it seemed to be a key in reproducing, if not this error, then some similar error.
	 ```
	 - protocol: TCP
	   port: 1080
	   name: tcp
	   allowedRoutes:
	     namespaces:
	       from: Selector
	       selector:
	         matchLabels:
	           kubernetes.io/metadata.name: default
	 ```
5. Scale the backend demo app to zero and up again and monitor requests to it

The behavior doesn't appear with the `zhaohuabing/gateway-dev:d3e3c6bf7` image it seems.

**Comment by zhaohuabing** [comment]
@@tskinner-oppfi Thanks for testing the fixes!  V1.2.2 will be released this week with the fix https://github.com/envoyproxy/gateway/pull/4754 .

**Comment by dromadaire54** [comment]
@zhaohuabing thanks for the fix no error during several days in our env.

**Comment by zhaohuabing** [comment]
Hello @dromadaire54 @tskinner-oppfi 
 If it's possible, could you please also try this image zhaohuabing/gateway-dev:96002955f ? It's built from this PR: https://github.com/envoyproxy/gateway/pull/4809

**Comment by tskinner-oppfi** [comment]
I've tested the image `zhaohuabing/gateway-dev:96002955f` in one of our EKS clusters and I still see errors although they are different than the 503s that were seen previously:

```
{"start_time":"2024-11-29T18:20:00.028Z","method":"-","x-envoy-origin-path":"-","protocol":"-","response_code":"0","response_flags":"UH","response_code_details":"-","connection_termination_details":"-","upstream_transport_failure_reason":"-","bytes_received":"0","bytes_sent":"0","duration":"0","x-envoy-upstream-service-time":"-","x-forwarded-for":"-","user-agent":"-","x-request-id":"-",":authority":"-","upstream_host":"-","upstream_cluster":"EmptyCluster","upstream_local_address":"-","downstream_local_address":"10.143.32.129:1080","downstream_remote_address":"172.28.6.182:65105","requested_server_name":"-","route_name":"-"}
```

**Comment by dromadaire54** [comment]
Hi,
We are facing 503 error code in the latest version `1.2.3` which fixed this issue. There's somebody else with the same situation ?

**Comment by ovaldi** [comment]
Recently, after upgrading to v1.2.0 and v1.2.1, we encountered the [503s](https://github.com/envoyproxy/gateway/issues/4685) issue.

And, after upgrading to v1.2.2, we encountered the [DPANIC](https://github.com/envoyproxy/gateway/issues/4805) issue.

So for v1.2.3, we decided to wait and see... 😂

> Hi, We are facing 503 error code in the latest version `1.2.3` which fixed this issue. There's somebody else with the same situation ?

**Comment by arkodg** [comment]
> Hi,
> We are facing 503 error code in the latest version `1.2.3` which fixed this issue. There's somebody else with the same situation ?

@dromadaire54 can you share the access logs o/p ? Im not sure how the fix that went into v1.2.3 can cause it (we only wait during init) , the issue @tskinner-oppfi is hitting is tied to a listener without an attached TCPRoute which looks like another issue (maybe configuration issue) and needs to be investigated separately
